### ISSUE TITLE
Consider deserializing serialized script blocks as such (remoting, background jobs)

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#11698]
Currently and historically, `[scriptblock]` instances are deserialized as `[string]` rather than as type `[scriptblock]`:

This leads to surprising behavior:

```powershell
# FAILS, because the script block deserializes to a string parameter
PS> pwsh -nop -c { param($sb) & $sb } -args { 'hi' }
&: The term ' 'hi' ' is not recognized as the name of a cmdlet, function, script file, or operable program.
```

```powershell
# FAILS, because the background job sees the script block as a string after deserialization.
PS> $sb = { 'hi' }; (Measure-Command $using:sb &) | Receive-Job -Wait -AutoRemoveJob
Measure-Command: Cannot bind parameter 'Expression'. Cannot convert the " 'hi' " value of type "System.String" to type "System.Management.Automation.ScriptBlock".
```

The _serialization_ part already contains all the required information - e.g. the above script block serializes as XML text `<SBK> 'hi' </SBK>`, but, surprisingly, during _deserialization_  `[scriptblock]::Create()` is _not_ called on this string.

Note: This simple plain-text serialization of the script block's content means that the `$using:`-scope feature for embedding values from the caller's scope doesn't work - references such as `$using:foo` are serialized as-is and would _fail_ on execution of the recreated script block in the caller's scope.

The historical design rationale for this behavior was given by @BrucePay  in https://github.com/PowerShell/PowerShell/issues/4218#issuecomment-314851921:

> Historically this is by design. Serializing scriptblocks with fidelity resulted in too many places where there was automatic code execution so to facilitate secure restricted runspaces, scriptblocks are always deserialized to strings.

To safely avoid this problem and to correct the surprising behavior, it  sounds like we'd need to simply make sure that PowerShell-_internally_ no accidental execution (not sure when and why that would happen; simply _parsing_ strings into script blocks seems to be allowed even in restricted runspaces).

From the _user_'s perspective, a script block should deserialize as such, and it is their responsibility, as always, to decide if and when to invoke it.

For old code that worked around the limitation, this change should have little to no impact, given how `[scriptblock]` instances stringify:

```powershell
# The following should be equivalent
[scriptblock]::create(" 'foo' ") # convert from string
[scriptblock]::create({ 'foo' }) # redundant conversion from script block, which old code may end up doing
```

Hypothetically, old code that tries to access `[string]` properties on a deserialized script block could break - hopefully, this scenario falls into [bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area).

**Comment by PaulHigin** [comment]
Remoting working group:
We could have a conversion that allows a deserialized script block to be re-parsed into a scriptblock.  But we will always just pass the script block string over the wire.  One reason for this is it is much lighter weight for performance reasons.  The other is that scriptblocks contain trust information that needs to be re-evaluated in the target session when the string is re-parsed into a script block.

**Comment by PaulHigin** [comment]
Remoting working group:
On further consideration, this is by design.  We don't want to be able to cast a deserialized scriptblock (or string) into a scriptblock, because this opens up security issues.  It is essentially a 'bad' constructor and is why scriptblock has a 'Create' function.

**Comment by mklement0** [comment]
> because this opens up security issues.

It doesn't open up security issues _per se_, because a script block is just a piece of code that must be _invoked explicitly_ in order to do anything (including potential harm).

By not deserializing to a script block, all you're doing is to create extra work for the user, by forcing them to call `[scriptblock]::Create()` themselves or - God forbid - `Invoke-Expression` - the need for which they won't expect, as demonstrated by the examples in the initial post.

**Comment by PaulHigin** [comment]
It is funny how pen testers can change your mind :).  The serialization system is used in remoting and we've had enough security reviews to know we don't want to do this.

**Comment by mklement0** [comment]
@PaulHigin, what are pen testers?

And how does "trust me, we've looked into this before" address the specific argument I've made?

**Comment by PaulHigin** [comment]
Penetration testers.  We will not implement automatic conversion of string to scriptblock via constructor as that can open up injection attacks as shown by our penetration testers.

**Comment by mklement0** [comment]
Thanks for clarifying the terminology.

So, how does serialization to a script block enable an injection attack if it requires an _explicit, deliberate act of invocation_ to execute the deserialized script block?

**Comment by mklement0** [comment]
Given how this conversation has unfolded, it is to be feared that this fiat decision stands, even though it is sole justification is the abstract claim that there is an injection-attack security concern, without supporting that claim or being willing to debate it.

Therefore, the following is likely only for the benefit of future readers (disclaimer: I am by no means a security expert):

From what I gather, an injection attack can mean one of two things:
  * Execution of (malicious) code where no execution of code is expected at all.
  * Execution of maliciously _modified_ code.

Neither scenario applies in the situation at hand:

* If what is _serialized_ as a script block (in type-specific) `<SBK>` elements, by mere stringification (the block's verbatim source code),
* then _deserializing_ it as such (via `[scriptblock]::Create()`) cannot possibly result in execution of unwanted code, because _constructing_ a script block doesn't result in its _execution_; it takes _explicit invocation_ (via `&`, `.Invoke()` methods) for actual execution to occur. 
  * That a stringified script block representation cannot fully capture all aspects of the original script block is incidental to this discussion; (partial) loss of type fidelity is inherent in serialization, whether applied cross-process locally or in remoting.

Note that concern about the serialized data itself getting maliciously modified is an entirely separate debate that is not specific to this use case.

To put it differently: If the user's intent in trying something like `pwsh -nop -c { param($sb) & $sb } -args { 'hi' }` is to execute the script block being passed, then they can make that happen even with the current limitation:
`pwsh -nop -c { param($sb) Invoke-Expression "& { $sb }" } -args { 'hi' }`

That it requires an _explicit_ act to recreate a serialized script block is nothing but an unexpected annoyance - it provides no security.

Conversely, if someone managed to tamper with the serialized data itself - an entire separate, fundamental problem - all bets would be off even with respect to the script block that _is_ directly executed,  `{ param($sb) & $sb }` / `{ param($sb) Invoke-Expression "& { $sb }" }{ param($sb) Invoke-Expression "& { $sb }" }`.

**Comment by dkattan** [comment]
I just ran into this issue myself. @PaulHigin Can you shed some light on how/why code could get executed if the ScriptBlock was deserialized automatically?

**Comment by oz53719** [comment]
Has the situation improved since this was declared "as designed"? I'm trying to move some PowerShell from an in-process session to an out-of-process session, but I can't invoke Invoke-Command -ScriptBlock $sb in the out-of-process session:

Cannot bind parameter 'ScriptBlock'. Cannot convert the "<script body>" value of type "System.String" to
type "System.Management.Automation.ScriptBlock".

It'd be great if the parameter value emerged from serialization/deserialization as a ScriptBlock. It'd be tolerable (in my current situation, at least) if Invoke-Command accepted -ScriptBlock <string> and internally called ScriptBlock.Create(). If neither of those are going to work, is there a documented workaround (with useful examples)?

**Comment by mklement0** [comment]
The workaround is much simpler: use `[scriptblock]::Create()` (e.g. `$sb = { 'hi' }; Start-Job { & ([scriptblock]::Create($using:sb)) } | Receive-Job -Wait -AutoRemove`); however, the point of this issue is that _no workaround should be needed_.

**Comment by suprak** [comment]
I agree. But given the stance and the digging that I did, the above is what unblocked my scenario.

**Comment by mklement0** [comment]
@suprak, I see.

Nice job on getting this to work without the need to modify the code in the script block itself (which is what my ad hoc workaround is based on), but how would you make this work in practice, given that you'd need access to the raw serialization data in order to apply your custom deserialization?
Or are you sending the raw serialization data _as a string_? 

To me it looks like _some_ workaround  that requires extra work on a _per-call_ basis is needed, or am I missing something?

**Comment by suprak** [comment]
Right, in my scenario I needed to serialize PS object state and deserialize it back reliably. Things like hashtables, and scriptblocks.
I would have just used Import/Export-Clixml except for the scriptblock deserializing back as strings and not scriptblocks.
After studying how clixml is generated, I realized that deserializing scriptblocks back into scriptblocks is supported just not exposed.
So what I pasted is the solution for my scenario.

**Comment by mklement0** [comment]
Makes sense; thanks for sharing.
